src/lib/momento/auth/credential.ex

```
@@ -0,0 +1,155 @@
+defmodule Momento.Auth.Credential do
+  require Joken
+
+  @moduledoc """
+  Handles decoding and managing Momento authentication credentials.
+  """
+
+  defstruct [
+    :control_endpoint,
+    :cache_endpoint,
+    :auth_token
+  ]
+
+  @opaque t :: %__MODULE__{
+            control_endpoint: String.t(),
+            cache_endpoint: String.t(),
+            auth_token: String.t()
+          }
+
+  @doc """
+  Fetches the given environment variable and parses it into a credential.
+
+  Returns the credential or raises an exception.
+
+  ## Examples
+
+      iex> Momento.Auth.Credential.parse_credential_from_env_var("MOMENTO_AUTH_TOKEN")
+      %Momento.Auth.Credential{}
+
+  """
+  @spec parse_credential_from_env_var(String.t()) :: t() | no_return()
+  def parse_credential_from_env_var(env_var) do
+    case System.get_env(env_var) do
+      nil -> raise "#{env_var} is not set"
+      token -> parse_credential(token)
+    end
+  end
+
+  @doc """
+  Parses the given string into a credential.
+
+  Returns the credential or raises an exception.
+
+  ## Examples
+
+      iex> valid_token = "valid_token" # This should be a valid Momento auth token.
+      iex> Momento.Auth.Credential.parse_credential(valid_token)
+      %Momento.Auth.Credential{}
+
+  """
+  @spec parse_credential(String.t()) :: t() | no_return()
+  def parse_credential(token) do
+    case decode_v1_token(token) do
+      {:error, _} ->
+        case decode_legacy_token(token) do
+          {:error, _} -> raise "Failed to decode auth token"
```

re: my comment below about the `!` - you are already raising here anyway, so you need a `!`, so I think we should just lean into that for now

---

I'll add `!` to the public methods. I specifically made those raise an exception instead of returning an error since that is unrecoverable and we throw in other SDKs when setting up.